### PR TITLE
register a terminating pod if containers running

### LINKED ISSUE #113289: Pod is stuck in terminating with "MountVolume.SetUp failed for volume ... not registered" error
### What happened?

Pod is stuck in terminating state when rapidly create and delete the pod, the kubelet log reported the volume setup error:
```
Oct 21 12:09:20 slave2 kubelet[4089]: E1021 12:09:20.197090    4089 nestedpendingoperations.go:348] Operation for "{volumeName:kubernetes.io/secret/dc79f107-6576-4d98-a120-23336154393e-tls-cert-vol podName:dc79f107-6576-4d98-a120-23336154393e nodeName:}" failed. No retries permitted until 2022-10-21 12:11:22.19703932 +0800 CST m=+323223.503182418 (durationBeforeRetry 2m2s). Error: MountVolume.SetUp failed for volume "tls-cert-vol" (UniqueName: "kubernetes.io/secret/dc79f107-6576-4d98-a120-23336154393e-tls-cert-vol") pod "vela-core-8494697c8c-xgrwt" (UID: "dc79f107-6576-4d98-a120-23336154393e") : object "cpaas-system"/"vela-core-admission" not registered
```

It has the same error with #105204, but the k8s 1.23.13 we used have fixed the issue with the PR #108756.

So the PR haven't totally fixed the issue. The cause may be if the pod is deleted when the volume manager is setting up the volume of the pod, the `syncPod` loop in kubelet won't add the reference count because it's in terminating state.

So the volume manager will still try to setup the volume of the pod but failed with the "not registed" error.

### What did you expect to happen?

The pod should be deleted without the volume setup error.

### How can we reproduce it (as minimally and precisely as possible)?

It's the possibel reproduce steps:
1. Create a pod with a configmap volume
2. After the pod is created, delete the pod immediately

Maybe the steps should be runned several times. 

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```
</details>
v1.23.13

### Cloud provider

<details>

</details>
None

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>
CentOS 7.3

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by olyazavr**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

When a pod is terminating, but its containers are still running (eg. it's running a prestop hook), and then kubelet restarts, it will re-sync but not the volumes, and that causes this error: `MountVolume.SetUp failed for volume \"kube-public-cert\" : object \"browsers-dashboard\"/\"kube-public-cert\" not registered"` which can only be resolved by force-deleting the pod

See this gist for logs: https://gist.github.com/olyazavr/f1980620e89dead80e72a23fcfe2014d

Basically we want [this logic that happens in SyncPod](https://github.com/kubernetes/kubernetes/blob/07b526c7ce6f53e5cf5e1c0c038aa30a26b95b4b/pkg/kubelet/kubelet.go#L1773-L1778) but in the [SyncTerminatingPod](https://github.com/kubernetes/kubernetes/blob/07b526c7ce6f53e5cf5e1c0c038aa30a26b95b4b/pkg/kubelet/pod_workers.go#L1275) flow

Testing: I have a pod with a prestop hook to sleep for 10m, and a secret mounted on it. I kill the pod, then restart kubelet. Without this change, it will error out with the mount volume error, and with it, that error does not appear and the pod terminates gracefully

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/113289

#### Does this PR introduce a user-facing change?

```release-note
Fixes a bug where if a pod is terminating and has a secret or configmap mounted and kubelet restarts, it would get MountVolume.SetUp errors and not complete terminating unless force-deleted 
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by HirazawaUi** [comment]
/ok-to-test
I guess we should write an e2e test for it?

File: test/e2e_node/terminating_pod_test.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+			ginkgo.By("wait for the pod to be terminating")
+			gomega.Eventually(ctx, func(ctx context.Context) error {
+				return verifyPodTerminating(ctx, f.ClientSet, podName, ns)
+			}, 2*time.Minute, time.Second*1).Should(gomega.BeNil())
+
+			ginkgo.By("wait for the pod to completely terminate")
+			gomega.Eventually(ctx, func(ctx context.Context) error {
+				return verifyPodDisappeared(ctx, f.ClientSet, podName, ns)
+			}, 2*time.Minute, time.Second*1).Should(gomega.BeNil())
```
Comment (HirazawaUi) [inline test/e2e_node/terminating_pod_test.go:60]:
```suggestion
		gomega.Eventually(ctx, func() bool {
			_, err := cl.CoreV1().Pods(ns).Get(ctx, name, metav1.GetOptions{})
			return apierrors.IsNotFound(err)
		}, 2*time.Minute, time.Second).Should(gomega.BeTrue())
```
It'll be easier.

File: test/e2e_node/terminating_pod_test.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+	admissionapi "k8s.io/pod-security-admission/api"
+
+	"github.com/onsi/ginkgo/v2"
+	"github.com/onsi/gomega"
+)
+
+var _ = SIGDescribe("TerminatingPod", func() {
```
Comment (HirazawaUi) [inline test/e2e_node/terminating_pod_test.go:39]: I think it should move to `test/e2e_node/terminate_pods_test.go`

**Comment by olyazavr** [comment]
Could I get some review on this? We still see this issue and it would be great to have a fix that doesn't require hard deleting the pods on the node to fix

**Comment by haircommander** [comment]
> @haircommander could I get some clarity as to why this was moved to the "archive it" section of the board? This problem still exists and this patch is currently running in our production environment to address it

sorry for the late reply, this got buried in my inbox. I archived it on the sig node ci board. it has some testing changes but it's primarily the concern of the main board (where it remains in the needs reviewer state).

**Pull request closed without merging**